### PR TITLE
Add bridge with cache

### PR DESCRIPTION
**Pull request by prin-r**

### CODE REVIEW

File: bridges/evm/contracts/Bridge.sol [hunk 1]
```diff
@@ -155,8 +155,8 @@ contract Bridge is IBridge, Ownable {
     /// Performs oracle state relay and oracle data verification in one go. The caller submits
     /// the encoded proof and receives back the decoded data, ready to be validated and used.
     /// @param _data The encoded data for oracle state relay and data verification.
-    function relayAndVerify(bytes calldata _data)
-        external
+    function relayAndVerify(bytes memory _data)
```
Comment (sorawit) [inline bridges/evm/contracts/Bridge.sol]: What the gas cost implication?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: CHANGELOG_UNRELEASED.md [hunk 2]
```diff
@@ -22,6 +22,10 @@
 
 ### Bridges
 
+- (feat) [\#2055](https://github.com/bandprotocol/bandchain/pull/2055) Implement BridgeWithCache to keep the latest response for any unique request packet
+- (bug) [\#2049](https://github.com/bandprotocol/bandchain/pull/2049) Fix signed data prefix not consistency bug.
+- (patch) [\#2041](https://github.com/bandprotocol/bandchain/pull/2041) Patch Multistore to new structure tree of 0.38.
```
Comment (taobun) [inline CHANGELOG_UNRELEASED.md]: Remove this change

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        require(ok, "FAIL_TO_RELAY_AND_VERIFY");
+
+        (RequestPacket memory req, ResponsePacket memory res) = abi.decode(
+            rawResult,
+            (RequestPacket, ResponsePacket)
+        );
+
+        require(
+            res.requestId > requestsCache[getRequestKey(req)].requestId,
+            "FAIL_REQUEST_SHOULD_NEWER_THAN_THE_LATEST"
```
Comment (smiled0g) [inline bridges/evm/contracts/BridgeWithCache.sol]: "FAIL_LATEST_REQUEST_SHOULD_BE_NEWEST"

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+        require(ok, "FAIL_TO_RELAY_AND_VERIFY");
+
+        (RequestPacket memory req, ResponsePacket memory res) = abi.decode(
+            rawResult,
+            (RequestPacket, ResponsePacket)
+        );
+
+        require(
+            res.requestId > requestsCache[getRequestKey(req)].requestId,
```
Comment (smiled0g) [inline bridges/evm/contracts/BridgeWithCache.sol]: use `res.resolveTime`

**Review by smiled0g — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            abi.encodeWithSignature("relayAndVerify(bytes)", _data)
+        );
+
+        require(ok, "FAIL_TO_RELAY_AND_VERIFY");
+
+        (RequestPacket memory req, ResponsePacket memory res) = abi.decode(
+            rawResult,
+            (RequestPacket, ResponsePacket)
+        );
+
```
Comment (taobun) [inline bridges/evm/contracts/BridgeWithCache.sol:51]: Can we store getRequestKey in a variable? remove duplicate call encode.

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 6]
```diff
@@ -0,0 +1,66 @@
+pragma solidity 0.5.14;
+pragma experimental ABIEncoderV2;
+
+import {Packets} from "./Packets.sol";
+import {Bridge} from "./Bridge.sol";
+
+/// @title BridgeWithCache <3 BandChain
+/// @author Band Protocol Team
+contract BridgeWithCache is Bridge {
```
Comment (taobun) [inline bridges/evm/contracts/BridgeWithCache.sol]: Inherit IBridgeCache too?

**Review by taobun — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    mapping(bytes32 => ResponsePacket) public requestsCache;
+
+    /// Initializes an oracle bridge to BandChain by pass the argument to the parent contract (Bridge.sol).
+    /// @param _validators The initial set of BandChain active validators.
+    constructor(ValidatorWithPower[] memory _validators)
+        public
+        Bridge(_validators)
+    {}
+
+    /// Create hash of a RequestPacket.
```
Comment (sorawit) [inline bridges/evm/contracts/BridgeWithCache.sol]: Returns the hash of

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    /// @param _request A tuple that represent RequestPacket struct.
+    function getRequestKey(RequestPacket memory _request)
```
Comment (sorawit) [inline bridges/evm/contracts/BridgeWithCache.sol:23]: represents

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        public
+        pure
+        returns (bytes32)
+    {
+        return keccak256(abi.encode(_request));
+    }
+
+    /// Query the ResponsePacket for a given RequestPacket.
```
Comment (sorawit) [inline bridges/evm/contracts/BridgeWithCache.sol]: Returns ... Reverts ...

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    /// Revert if can't find the related response in the mapping.
+    /// @param _request A tuple that represent RequestPacket struct.
+    function getLatestResponse(RequestPacket memory _request)
```
Comment (sorawit) [inline bridges/evm/contracts/BridgeWithCache.sol:34]: represents

File: bridges/evm/contracts/IBridgeCache.sol [hunk 11]
```diff
@@ -0,0 +1,13 @@
+pragma solidity 0.5.14;
+pragma experimental ABIEncoderV2;
+
+import {IBridge} from "./Bridge.sol";
+
+interface IBridgeCache {
+    /// Query the ResponsePacket for a given RequestPacket.
```
Comment (sorawit) [inline bridges/evm/contracts/IBridgeCache.sol]: ditto

**Review by sorawit — CHANGES_REQUESTED** [review]

File: bridges/evm/contracts/BridgeWithCache.sol [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        require(res.requestId != 0, "RESPONSE_NOT_FOUND");
+
+        return res;
+    }
+
+    /// Performs oracle state relay and oracle data verification in one go.
+    /// After that, the results will be recorded to the state by using the hash of RequestPacket as key.
+    /// @param _data The encoded data for oracle state relay and data verification.
+    function relay(bytes calldata _data) external {
+        (bool ok, bytes memory rawResult) = address(this).call(
```
Comment (sorawit) [inline bridges/evm/contracts/BridgeWithCache.sol]: We can just do `this.relayAndVerify`

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Review by smiled0g — APPROVED** [review]

**Pull request merged**